modules/core/src/ovx.cpp

```diff
@@ -14,6 +14,72 @@
 namespace cv
 {
 
+namespace ovx
+{
+#ifdef HAVE_OPENVX
+class TLOVXContextContainer : protected TLSDataContainer
+{
+public:
+    inline TLOVXContextContainer() {}
+    inline ~TLOVXContextContainer() { release(); } // Release key and delete associated data
+    inline ivx::Context* get() const { return (ivx::Context*)getData(); } // Get data assosiated with key
+
+                                                    // Get data from all threads
+    inline void gather(std::vector<ivx::Context*> &data) const
+    {
+        std::vector<void*> &dataVoid = reinterpret_cast<std::vector<void*>&>(data);
+        gatherData(dataVoid);
+    }
+
+private:
+    virtual void* createDataInstance() const { return new ivx::Context(ivx::Context::create()); }
+    virtual void  deleteDataInstance(void* pData) const { delete (ivx::Context*)pData; }
+    // Disable copy operations
+    TLOVXContextContainer(TLOVXContextContainer &) {}
+    TLOVXContextContainer& operator =(const TLOVXContextContainer &) { return *this; }
+};
+
+class TLOVXContextSingleton
+{
+public:
+    inline TLOVXContextSingleton() {}
+    inline ~TLOVXContextSingleton()
+    {
+        if (instance != NULL)
+        {
+            cv::AutoLock lock(cv::getInitializationMutex());
+            if (instance != NULL)
+            {
+                delete instance;
+                instance = NULL;
+            }
+        }
+    }
+    inline TLOVXContextContainer& getTLContextContainer()
+    {
+        if (instance == NULL)
```

Comment:
There is data race from `~TLOVXContextSingleton()` on statement `instance = NULL`.